jax/interpreters/xla.py

```diff
@@ -374,26 +386,31 @@ def jaxpr_subcomp(c, jaxpr, backend, axis_env, consts, name_stack, *args):
   else:
     platform = backend
 
-  def read(v):
-    if type(v) is Literal:
-      return xb.constant(c, canonicalize_dtype(v.val))
-    else:
-      return env[v]
+  def read(vals):
```

Comment:
I think the argument should be called `vars` rather than `vals`.

But I'm not a fan of having the `read` function take in a list of variables; it's unlike all our other jaxpr interpreters, and even unlike other functions here (like `aval`). What's the downside to keeping its signature so that it reads the value for a single variable (and that value is a sequence)? (Same for `write`.)

I think implementing it in terms of generators is a bit too tricky and obscures what's going on.

Reply:
I did it this way becuase every time `read` is used, it's used as `map(read, vals)`. Options are either to write the generator logic once here, or replicate it each time `read()` is used; the former seemed cleaner to me.

Reply:
Another option would be to add a `flatmap()` utility that does the generator logic, and call `read()` as `flatmap(read, vals)`. Would you find that preferable?

Reply:
(I know we've talked about this before, thanks for your patience.)

Yes, I like the idea of `flatmap(read, vars)` better than the current code, but actually I think it'd be better not to flatten at all. The flattening obscures information, which then `write` needs to recover (using `aval._num_buffers` etc).

IIUC flattening is just one way to preserve some of the existing calling convention for translation rules, namely
1. allowing existing translation rules, called like `rule(c, *in_nodes, **eqn.params)`, to expect each splatted in_node to be an XlaOp, rather than, say, a single-element tuple (if `in_nodes` were a list-of-lists), and
2. allowing us to keep the current `jaxpr_subcomp` signature where `consts` and `args` are sequences of XlaOps rather than lists-of-lists.

An alternative might be not to flatten, i.e. to have the type

```python
def read(v: Union[Var, Literal]) -> Union[XlaOp, Tuple[XlaOp]]:
  ...

...
in_nodes : List[Union[XlaOp, Tuple[XlaOp]] = map(read, eqn.invars)
```

Then we still get (1) because unless we're dealing with multi-buffer avals we know `read(var) : XlaOp`, and hence `in_nodes : List[XlaOp]` just like before. And I think (2) won't be too hard to handle directly (unless I'm forgetting something).

Are there more challenges/issues to sort out, other than (1) and (2)?

Reply:
If we avoid flattening, we run into the issue of re-introducing XLA tuples into JAX, which is problematic for other reasons. I have a working version of that here: https://github.com/jakevdp/jax/commit/675fea513fb6993132679eccbe702eef6f19223d which we'd previously ruled-out.

Reply:
I don't mean to suggest we change anything about what XLA sees; I'm just talking about bookkeeping on the Python side. That is, we still create the exact same XlaOps, we just keep them organized in a `List[Union[XlaOp, Tuple[XlaOp]]` instead of in a flat `List[XlaOp]`.

Reply:
Let's chat about this offline – I'm not certain what's different between the approach you're proposing and the approach I implemented in the commit linked above.

Reply:
Doing the flatmap approach for now, perhaps we can followup by updating translation rules to handle nested lists of buffers.

Reply:
(We chatted offline!) SGTM, land and iterate!